### PR TITLE
Adds fake Esri client

### PR DESCRIPTION
**Pull request by seanmalbert**
### Description

Per the [Address Correction and Service Area Validation Testing Strategy](https://docs.google.com/document/d/1_rXidQeZ2bL0rg5hDEPJmFazVsdQjsuPnXZ1xINXDAU/edit#) doc, this adds a fake esri client to use for dev and testing. This does not cover adding a mock service to simulate external IO.

To test in development you will need to enable the address correction and service area validation features. In admin, you will also have to enable address correction for an address question and add service area eligiblity on the block.

When you get to the applicant address question, there are four options you can enter for the address value. The other field values do not matter.
1. `Legit Address` - this will provide three suggestions for tesing service area validation:
    1. `Address In Area`
    2. `Address With No Service Area Features`
    3. `Address Not In Area`
2. `Bogus Address` - Use this for testing address that won't return any suggestions. This is likely not to happen in the real world, but you'll see that this isn't handled properly at the moment (and won't be handled in this PR).
3. `Error Address` - Use this to test what would happen if the external Esri service were to return a non 200 status. The end result is the same as `Bogus Address`.
4. Any other address will have the same effect as Bogus and Error.

### CODE REVIEW

File: server/app/services/geo/esri/FakeEsriClient.java
Comment (avrittrohwer) [inline server/app/services/geo/esri/FakeEsriClient.java:1]: Could you add the known address input values in the comments here?

File: server/app/services/applicant/ApplicantService.java
Comment (avrittrohwer) [reply server/app/services/applicant/ApplicantService.java:112]: I did some looking into this and I think this can be achieved by adding a new Esri module: https://www.playframework.com/documentation/2.8.x/JavaDependencyInjection#Programmatic-bindings

For example, the file storage provider strategy implementation is provided based on a value in the config: https://github.com/civiform/civiform/blob/main/server/app/modules/CloudStorageModule.java.  We could provider the FakeEsriClient if DeploymentType.isDevOrStaging() is true

Comment (gwendolyngoetz) [reply server/app/services/applicant/ApplicantService.java:112]: Building on what Avritt said, rather than inject both EsriClient and FakeEsriClient I think we should do one of the following:
* Create a [custom Guice module](https://www.playframework.com/documentation/2.8.x/JavaDependencyInjection#Programmatic-bindings) to allow the dependency injection framework to inject the appropriate type (recommended). This will allow just having to inject the one interface.
* Create a factory that can determine the appropriate object to use, and inject the factory.

**Review by gwendolyngoetz — CHANGES_REQUESTED** [review]

File: server/app/services/applicant/ApplicantService.java
Comment (seanmalbert) [reply server/app/services/applicant/ApplicantService.java]: I didn't end up using DeploymentType, since we want to be able to use the real esri client in staging.

File: server/app/services/geo/esri/FakeEsriClient.java
Comment (seanmalbert) [reply server/app/services/geo/esri/FakeEsriClient.java:1]: added options in comment for `fetchAddressSuggestions`.

File: server/app/services/applicant/ApplicantService.java
Comment (seanmalbert) [reply server/app/services/applicant/ApplicantService.java]: Thanks for the great suggestion @avrittrohwer and @gwendolyngoetz! Added an EsriModule and EsriClient is now an abstract class. Went with a class instead of an interface for EsriClient, since `getAddressSuggestions` and `getServiceAreaInclusionGroup` methods are the same between RealEsriClient and FakeEsriClient.

**Comment by seanmalbert** [comment]
Latest main has bugs, so waiting until after that's fixed to resolve merge conflict.

File: server/app/services/applicant/ApplicantService.java
Comment (nb1701) [inline server/app/services/applicant/ApplicantService.java:1265]: I think the formatter mangled this comment

File: server/app/services/geo/esri/EsriClient.java
Comment (gwendolyngoetz) [inline server/app/services/geo/esri/EsriClient.java:78]: Nitpick: Remove `this.`

File: server/app/services/geo/esri/EsriClient.java
Comment (gwendolyngoetz) [inline server/app/services/geo/esri/EsriClient.java:80]: I think the logged message should still use `EsriClient` and not `RealEsriClient` since it's within the EsriClient. If this is specific to the Real client we should make it abstract and move this into there.

File: server/app/services/geo/esri/EsriClient.java
Comment (gwendolyngoetz) [inline server/app/services/geo/esri/EsriClient.java:181]: Same as above comments on the logger error message about `RealEsriClient`

File: server/app/services/geo/esri/EsriClient.java
Comment (seanmalbert) [reply server/app/services/geo/esri/EsriClient.java]: updated

File: server/app/services/geo/esri/EsriClient.java
Comment (seanmalbert) [reply server/app/services/geo/esri/EsriClient.java]: Thanks for catching this. I think I had an iteration where this was just in the Real one and didn't change this back.

File: server/app/services/geo/esri/EsriClient.java
Comment (seanmalbert) [reply server/app/services/geo/esri/EsriClient.java]: updated

**Review by gwendolyngoetz — APPROVED** [review]

File: server/app/modules/EsriModule.java
Comment (bion) [inline server/app/modules/EsriModule.java:12]: **suggestion (blocking):** please add javadoc

File: server/app/modules/EsriModule.java
Comment (bion) [inline server/app/modules/EsriModule.java:27]: **suggestion (blocking):** extract to constant

File: server/app/modules/EsriModule.java
Comment (bion) [inline server/app/modules/EsriModule.java:33]: **suggestion (blocking):** log which one is being used

File: server/app/modules/EsriModule.java
Comment (bion) [inline server/app/modules/EsriModule.java:39]: **suggestion (blocking):** 
```suggestion
      throw new RuntimeException(String.format("Failed to load esri client class: %s", className), e);
```

File: server/app/services/geo/esri/FakeEsriClient.java
Comment (bion) [inline server/app/services/geo/esri/FakeEsriClient.java:23]: **nitpick:** 
```suggestion
public final class FakeEsriClient extends EsriClient {
```

File: server/app/services/geo/esri/RealEsriClient.java
Comment (bion) [inline server/app/services/geo/esri/RealEsriClient.java:37]: **nitpick:** 
```suggestion
public final class RealEsriClient extends EsriClient implements WSBodyReadables, WSBodyWritables {
```

**Review by bion — CHANGES_REQUESTED** [review]

File: server/app/services/geo/esri/FakeEsriClient.java
Comment (bion) [inline server/app/services/geo/esri/FakeEsriClient.java:29]: **issue (blocking):** Ensure these members won't be reused between requests.

By default, Play uses dependency injection to instantiate an instance of the controller needed to handle a request, and then reuses that instance for subsequent requests. Dependencies of that controller (and therefore their dependencies) will similarly be reused. There are two options for changing this behavior, one is to inject a `Provider<YouClass>` and call `get` on it to force creation of a new instance. The other is to prefix the route with an `@` in the routes file, forcing Play to inject a `Provider<YourController>` in its generated router class.

More info here: https://www.playframework.com/documentation/2.8.x/JavaDependencyInjection#Dependency-injecting-controllers

File: server/app/services/geo/esri/FakeEsriClient.java
Comment (bion) [inline server/app/services/geo/esri/FakeEsriClient.java:30]: **nitpick:** 
```suggestion
  private final Logger LOGGER = LoggerFactory.getLogger(this.getClass());
```

File: server/app/services/geo/esri/FakeEsriClient.java
Comment (bion) [inline server/app/services/geo/esri/FakeEsriClient.java:92]: **suggestion (blocking):** at least log an error here, consider throwing an exception as well to signal the error state to the caller

**Review by bion — CHANGES_REQUESTED** [review]

File: server/app/services/geo/esri/FakeEsriClient.java
Comment (nb1701) [reply server/app/services/geo/esri/FakeEsriClient.java:29]: If I'm reading this correctly (and I may not be), I think it's okay for these to be reused.  They will contain the contents of the json files that are used for the mock data, and it's just a caching mechanism for that.  None of that data will change.

**Review by bion — APPROVED** [review]

**Pull request merged**
